src/classes/shex-form-model.ts

```diff
@@ -0,0 +1,266 @@
+
+    const writer = new Writer({
+      prefixes: { '': IRI_this, ui: NS_UI, dc: NS_DC },
+      listHeads: graph.sequesterLists()
+    })
+    writer.addQuads(graph.getQuads())
+    let formModel
+    writer.end((error, result) => (formModel = result))
+
+    return formModel
+  }
+  /**
+   * Find label expression into ShEx
+   * @param shexpr
+   */
+  findTitle(shexpr: any) {
+    return (shexpr.annotations || []).find((a: any) => a.predicate === this.iriDctitle)
+  }
+
+  /**
+   * Define shape type into formModel
+   * @param iri
+   */
+  localName(iri: string) {
+    const { meta } = this
+
+    if (iri.startsWith('_:')) {
+      return iri
+    }
+    let prefix = Object.keys(meta.prefixes).find(p => iri.startsWith(meta.prefixes[p]))
+
+    if (prefix) {
+      return prefix + `:${iri.substr(meta[prefix].length)}`
+    }
+    return `<${iri.startsWith(meta.base) ? iri.substr(meta.base.length) : iri}>`
+  }
+  /**
+   *
+   * @param shapeExpr
+   */
+  derefShapeExpression(shapeExpr: any) {
+    if (typeof shapeExpr !== 'string') {
+      return shapeExpr
+    }
+    const ret = this.findShapeExpression(shapeExpr)
+    if (!ret) {
+      throw Error(`unable to find shape expression "${shapeExpr}" in 
+            ${this.schema.shapes.map((se: any) => se.id).join('\n')}`)
+    }
+    return ret
+  }
+
+  /**
+   * Find shape expression with given name in schema.
+   * returns: corresponding shape expression or undefined
+   * @param {string} goal expression to find
+   */
+  findShapeExpression(goal: string) {
+    return this.schema.shapes.find((se: any) => se.id === goal)
+  }
+  /**
+   * Run hover shape and create the object with turtle format
+   * @param shape
+   * @param formTerm
+   * @param path
+   * @param namedNode
+   * @param literal
+   * @param blankNode
+   */
+  walkShape(shape: any, formTerm: any, path: string, namedNode: any, literal: any, blankNode: any) {
+    try {
+      const { graph } = this
+      const sanitizedPath = path.replace(/[^A-Za-z_-]/g, '_')
+      const label = this.findTitle(shape)
+
+      graph.addQuad(formTerm, namedNode(`${NS_RDF}type`), namedNode(`${NS_UI}Form`))
+
+      if (label) {
+        graph.addQuad(formTerm, namedNode(this.iriDctitle), literal(label.object.value))
+      }
+
+      let currentShape = shape
+
+      if (!('expression' in shape) || shape.expression.type !== 'EachOf') {
+        currentShape = {
+          ...currentShape,
+          expression: { expressions: [currentShape.expression] }
+        }
+      }
+
+      // The UI vocabulary accepts only lists of atoms.
+      let parts = new ListObject(formTerm, namedNode(`${NS_UI}parts`), graph, this.termFactory)
+
+      if (currentShape && currentShape.expression && currentShape.expression.expressions) {
+        currentShape.expression.expressions.forEach((te: any, i: any) => {
+          const tePath = `${path}/[${i}]`
+          if (te.type !== 'TripleConstraint') {
```

Comment:
Another magic string (there are more on this page, I won't add a comment for each one)

Reply:
Done